<h1>B09705039_劉惟恩_HW05

In [1]:
# 載入所需函示庫
import pandas as pd
import numpy as np
import scipy.stats as stats
import statsmodels.api as sm
import statsmodels.stats.api as sms
import statsmodels.formula.api as smf
import statsmodels.stats.multicomp as smm
import statsmodels
import scipy
from matplotlib import pyplot as plt
%matplotlib inline
# 設定圖形大小; DPI越大圖越大
plt.rcParams["figure.dpi"] = 120

<h2> Problem 15.13

In [2]:
# 讀取資料集
df1 = pd.read_excel('Xr15-13.xlsx')

print('Head of dataset:')
display(df1.head())

print("Tail of dataset:")
display(df1.tail())

Head of dataset:


Car type
0         3
1         2
2         2
3         3
4         2

Tail of dataset:


Car type
192         2
193         1
194         3
195         3
196         4

Hypotheses:<br>
Define $a_i$ as the proportion of car type $i$ of license records. <br>
Define $p_i$ as the proprotion of car type $i$ involved in accident. <br>

$H_0$: $p_i$ = $a_i$ <br>
$H_1$: At least one $p_i \neq a_i$

In [3]:
df2 = df1["Car type"].value_counts()

nobs = df1.count()["Car type"]
alpha = 0.05
prob_e = np.array([0.15, 0.25, 0.40, 0.20])
freq_e = nobs * prob_e
freq_o = np.array([df2[1], df2[2], df2[3], df2[4]])

if np.sum(freq_e < 5) > 0:
    print("Rule of five is not met. ")
else:
    stat, pvalue = stats.chisquare(freq_o, freq_e)
    print("Chi-squared test: statistics = %0.4f, p-value = %0.4f" % (stat, pvalue))

df3 = freq_o.shape[0]-1
crit_value = stats.chi2.ppf(1 - alpha, df3)
print("Critical value = %0.4f (defree of freedom = %d)" % (crit_value, df3))

Chi-squared test: statistics = 6.0000, p-value = 0.1116
Critical value = 7.8147 (defree of freedom = 3)


Since p-value = 0.1116 > 0.05, there is no sufficient evidence to reject $H_0$.<br>
Thus, we cannot infer that certain sizes of cars are involved in a higher than expected percentage of accidents.

<h2> Problem 15.19

In [4]:
# 讀取資料集
df1 = pd.read_excel('Xr15-19.xlsx')

print('Head of dataset:')
display(df1.head())

print("Tail of dataset:")
display(df1.tail())

Head of dataset:


Month  Number of tornadoes
0   January                   26
1  February                    2
2     March                   13
3     April                  185
4       May                  414

Tail of dataset:


Month  Number of tornadoes
7      August                   44
8   September                   24
9     October                   57
10   November                   95
11   December                   93

Hypotheses:<br>
Define $a_i$ as the proportion of tornadoes struck the United States between 2000 and 2014 in month $i$. (1 as Jan, 2 as Feb ......) <br>
Define $p_i$ as the proportion of tornadoes struck the United States in 2015 in month $i$. <br>

$H_0$: $p_i$ = $a_i$ <br>
$H_1$: At least one $p_i \neq a_i$

The sum of $a_i$ is not 100%, we add 0.1% to month 2 and 7 ,since it has .0% which implies a rounding from the original data.  

In [5]:
nobs = df1["Number of tornadoes"].sum()
alpha = 0.05
prob_e = np.array([0.024, 0.031, 0.067, 0.169, 0.212, 0.174, 0.081, 0.057, 0.059, 0.049, 0.049, 0.028])
freq_e = nobs * prob_e
freq_o = np.array(df1["Number of tornadoes"])

if np.sum(freq_e < 5) > 0:
    print("Rule of five is not met. ")
else:
    stat, pvalue = stats.chisquare(freq_o, freq_e)
    print("Chi-squared test: statistics = %0.4f, p-value = %0.4f" % (stat, pvalue))

df3 = freq_o.shape[0]-1
crit_value = stats.chi2.ppf(1 - alpha, df3)
print("Critical value = %0.4f (defree of freedom = %d)" % (crit_value, df3))

Chi-squared test: statistics = 348.0236, p-value = 0.0000
Critical value = 19.6751 (defree of freedom = 11)


Since p-value = 0.0000... < 0.05, there is sufficient evidence to reject $H_0$, in favor of $H_1$.<br>
Thus, we can infer that the percentages have changed.

<h2> Problem 15.38

In [6]:
# 讀取資料集
df1 = pd.read_excel('Xr15-38.xlsx')

print('Head of dataset:')
display(df1.head())

print("Tail of dataset:")
display(df1.tail())

Head of dataset:


Education  Smoker
0          2       1
1          2       1
2          4       2
3          2       1
4          3       2

Tail of dataset:


Education  Smoker
995          2       2
996          2       2
997          2       1
998          2       1
999          3       1

Hypotheses:<br>
$H_0$: The Educational level is independent of continued to smoke 1 year later. <br>
$H_1$: The Educational level is dependent of continued to smoke 1 year later.

In [7]:
# 列出Contingency Table
cont1 = pd.crosstab(df1["Education"], df1["Smoker"])
print("Contingency Table:")
display(cont1)

Contingency Table:


Smoker       1    2
Education          
1           34   23
2          251  212
3          159  248
4           16   57

In [8]:
chi2, p, dof, ex = stats.chi2_contingency(cont1, correction=False)
print("Chi-squared Statistics = %0.4f" % chi2)
print("Degree of freedom = %0.4f" % dof)
print("p-value = %0.4f" % p)
print("Expected Frequency:")
print(ex)

Chi-squared Statistics = 41.7645
Degree of freedom = 3.0000
p-value = 0.0000
Expected Frequency:
[[ 26.22  30.78]
 [212.98 250.02]
 [187.22 219.78]
 [ 33.58  39.42]]


All the expected values satisfies rule of five.<br>

Since p-value = 0.0000... < 0.05, there is sufficient evidence to reject $H_0$, in favor of $H_1$.<br>
Thus, we can infer that the amount of education is a factor in determining whether a smoker will quit.

<h2>Problem 15.43

In [9]:
# 讀取資料集
df1 = pd.read_excel('Xr15-43.xlsx')

print('Head of dataset:')
display(df1.head())

print("Tail of dataset:")
display(df1.tail())

Head of dataset:


Year 2000  Year 2005  Year 2010
0        3.0          1        4.0
1        4.0          4        4.0
2        1.0          3        2.0
3        1.0          2        4.0
4        4.0          1        1.0

Tail of dataset:


Year 2000  Year 2005  Year 2010
412        NaN          1        NaN
413        NaN          4        NaN
414        NaN          4        NaN
415        NaN          4        NaN
416        NaN          4        NaN

Year 2000 and Year 2005<br>

Hypotheses:<br>
$H_0$: Year 2000 is independent of Year 2005. <br>
$H_1$: Year 2000 is dependent of Year 2005. <br>

Combine column 2 and 3.

In [10]:
# 列出Contingency Table
cont1 = pd.crosstab(df1["Year 2000"], df1["Year 2005"])
print("Contingency Table:")
cont1[2] = cont1[2] + cont1[3]
cont1.pop(3)
display(cont1)

chi2, p, dof, ex = stats.chi2_contingency(cont1, correction=False)
print("Chi-squared Statistics = %0.4f" % chi2)
print("Degree of freedom = %0.4f" % dof)
print("p-value = %0.4f" % p)
print("Expected Frequency:")
print(ex)

Contingency Table:


Year 2005   1   2   4
Year 2000            
1.0        71  32  58
2.0        13   8  15
3.0        23   6  24
4.0        67  29  63

Chi-squared Statistics = 3.3636
Degree of freedom = 6.0000
p-value = 0.7620
Expected Frequency:
[[68.49388753 29.52322738 62.98288509]
 [15.31540342  6.60146699 14.08312958]
 [22.54767726  9.71882641 20.73349633]
 [67.64303178 29.15647922 62.200489  ]]


Since p-value = 0.7620 > 0.05, there is no sufficient evidence to reject $H_0$, in favor of $H_1$.<br>

Year 2000 and Year 2010<br>

Hypotheses:<br>
$H_0$: Year 2000 is independent of Year 2010. <br>
$H_1$: Year 2000 is dependent of Year 2010. <br>

Combine column 2, 3 and 4.

In [11]:
# 列出Contingency Table
cont1 = pd.crosstab(df1["Year 2000"], df1["Year 2010"])
print("Contingency Table:")
cont1[2] = cont1[2] + cont1[3]
cont1.pop(3)
cont1[2] = cont1[2] + cont1[4]
cont1.pop(4)
display(cont1)

chi2, p, dof, ex = stats.chi2_contingency(cont1, correction=False)
print("Chi-squared Statistics = %0.4f" % chi2)
print("Degree of freedom = %0.4f" % dof)
print("p-value = %0.4f" % p)
print("Expected Frequency:")
print(ex)

Contingency Table:


Year 2010  1.0  2.0
Year 2000          
1.0         50   75
2.0         11   17
3.0         19   23
4.0         51   62

Chi-squared Statistics = 0.8871
Degree of freedom = 3.0000
p-value = 0.8285
Expected Frequency:
[[53.16558442 71.83441558]
 [11.90909091 16.09090909]
 [17.86363636 24.13636364]
 [48.06168831 64.93831169]]


Since p-value = 0.8285 > 0.05, there is no sufficient evidence to reject $H_0$, in favor of $H_1$.<br>

Year 2005 and Year 2010<br>

Hypotheses:<br>
$H_0$: Year 2005 is independent of Year 2010. <br>
$H_1$: Year 2005 is dependent of Year 2010. <br>

Combine column 2, 3 and 4.

In [12]:
# 列出Contingency Table
cont1 = pd.crosstab(df1["Year 2005"], df1["Year 2010"])
print("Contingency Table:")
cont1[2] = cont1[2] + cont1[3]
cont1.pop(3)
cont1[2] = cont1[2] + cont1[4]
cont1.pop(4)
display(cont1)

chi2, p, dof, ex = stats.chi2_contingency(cont1, correction=False)
print("Chi-squared Statistics = %0.4f" % chi2)
print("Degree of freedom = %0.4f" % dof)
print("p-value = %0.4f" % p)
print("Expected Frequency:")
print(ex)

Contingency Table:


Year 2010  1.0  2.0
Year 2005          
1           48   77
2           11   17
3           17   10
4           55   73

Chi-squared Statistics = 5.6149
Degree of freedom = 3.0000
p-value = 0.1319
Expected Frequency:
[[53.16558442 71.83441558]
 [11.90909091 16.09090909]
 [11.48376623 15.51623377]
 [54.44155844 73.55844156]]


Since p-value = 0.1319 > 0.05, there is no sufficient evidence to reject $H_0$, in favor of $H_1$.<br>


Thus, we can infer that the frequency of the use of weapons in robberies differed over the three years.

<h2> Problem 15.80

Hypotheses:<br>
$H_0$: Data are normaly distributed. <br>
$H_1$: Data are not normaly distributed. <br>

In [13]:
total_ob = 50
freq_i = np.array([6, 27, 14, 3])
exp_i = np.array([(total_ob / 4)] * 4)

crit_v = scipy.stats.chi2.ppf(1 - 0.1, 4 - 3)
print("critical value =", crit_v)

x_2 = ((freq_i - exp_i) ** 2) / (total_ob / 4)
x_2 = np.sum(x_2)
print("x^2 =", x_2)

critical value = 2.705543454095404
x^2 = 27.599999999999998


x^2 > critical value, we reject $H_0$ in favor of $H_1$.<br>
We can infer that the data are not normal

<h2> Problem 15.85

In [14]:
# 讀取資料集
df1 = pd.read_excel('Xr13-115.xlsx')

print('Head of dataset:')
display(df1.head())

print("Tail of dataset:")
display(df1.tail())

Head of dataset:


Method 1  Method 2
0       8.8       9.2
1       9.6       9.4
2       8.4       8.9
3       9.0       9.6
4       8.3       9.7

Tail of dataset:


Method 1  Method 2
5       9.2       8.4
6       9.0       8.8
7       8.7       8.9
8       8.5       9.0
9       9.4       9.7

Hypotheses:<br>
$H_0$: Data are normaly distributed. <br>
$H_1$: Data are not normaly distributed. <br>

We cannot satisfy the rule of five, because there isn't enough interval to test it.

In [15]:
diff = df1["Method 1"] - df1["Method 2"]
mean = diff.mean()
std = diff.std()
print(diff)

z1 = -1
z2 = 0
z3 = 1
x1 = z1 * std + mean
x2 = z2 * std + mean
x3 = z3 * std + mean
print("x1 =", x1, "x2 =", x2, "x3 =", x3, "\n")

total_ob = 10
freq_i = np.array([1, 4, 4, 1])
exp_i = np.array([(total_ob / 4)] * 4)

crit_v = scipy.stats.chi2.ppf(1 - 0.1, 4 - 3)
print("critical value =", crit_v)

x_2 = ((freq_i - exp_i) ** 2) / (total_ob / 4)
x_2 = np.sum(x_2)
print("x^2 =", x_2)

0   -0.4
1    0.2
2   -0.5
3   -0.6
4   -1.4
5    0.8
6    0.2
7   -0.2
8   -0.5
9   -0.3
dtype: float64
x1 = -0.8569885480616761 x2 = -0.2699999999999999 x3 = 0.3169885480616763 

critical value = 2.705543454095404
x^2 = 3.6


x^2 > critical value, we reject $H_0$ in favor of $H_1$.<br>
We still cannot infer that the data are not normal, because it violates the rule of five.
But it is likely that the normality requirement is violated.